cli/src/repl.rs

```diff
@@ -1,47 +1,230 @@
-use crate::cli::handle_source;
-use crate::report::print_flush;
+use crate::cli::{display_exprs, display_tokens, execute, Configuration};
+use crate::display_bytecode::display_bytecode;
+use crate::formatted_diagnostic::render_diagnostic;
+use crate::formatted_parse_error::render_parse_error;
+use analyzer::diagnostic::Diagnostic;
+use analyzer::engine::Engine;
+use analyzer::importer::ASTImporter;
+use analyzer::imports::Imports;
+use analyzer::name::Name;
+use analyzer::relations::{Definition, Relations, SourceId};
+use analyzer::steps::typing::apply_types;
+use analyzer::types::engine::CodeEntry;
+use ast::group::Block;
+use ast::Expr;
+use compiler::compile;
 use context::source::{OwnedSource, Source};
+use lexer::lexer::lex;
 use parser::parse;
-use std::io;
-use std::io::BufRead;
-use std::io::Write;
+use rustyline::config::Configurer;
+use rustyline::error::ReadlineError;
+use rustyline::history::DefaultHistory;
+use rustyline::{
+    Cmd, ColorMode, DefaultEditor, Editor, Event, EventHandler, KeyCode, KeyEvent, Modifiers,
+};
+use std::collections::HashMap;
+use std::process::exit;
+
+type REPLEditor = Editor<(), DefaultHistory>;
+
+#[derive(Default)]
+struct REPLImporter<'a> {
+    #[allow(clippy::vec_box)]
+    // Box is used to ensure that the reference behind is still valid after vector's realloc
+    sources: Vec<Box<OwnedSource>>,
```

Comment:
This is not necessary to box the source, as only the string pointers will be exposed and not the `Source` itself.

Reply:
if the vector is reallocated, the OwnedSource will be moved, `Source` points to the OwnedSource, thus if moved the references will be invalid